### DISCUSSION TITLE
What Du you think about a Video-Synthesizer Implementation?

### DISCUSSION THREAD

**Discussion by flowerhorse** [brunoherbelin/vimix#78] [Ideas]
for most advanced working possibilities? Maybe connected with the 'Filter Pad' I wrote. about 
It would also not be bad to be able to redirect the output to a smartphone as well. 
What do you think about that?

**Comment by brunoherbelin** [comment]
its not really a video-synthesizer, but you can do  (live) [coding in GLSL like in ShaderToy](https://github.com/brunoherbelin/vimix/wiki/Filters-and-ShaderToy#custom-filter-with-shadertoy-coding) . There are also all the gstreamer [frei0r](https://frei0r.dyne.org/) plugins available as custom generator source. 

In practice, i would rather prefer to create a bridge (e.g. a shared memory readout, or a network streaming) with another software. I quickly checked online and there doesn't seems to be many open source video synth software. Maybe this one ? https://github.com/daviderovell0/bzzzbz. Or let me know if you use a software that can stream its output, we could add a way to create an input source.

And yes, for streaming to a smartphone, its [done with SRT ](https://github.com/brunoherbelin/vimix/wiki/SRT-stream-I-O#srt-player-on-a-phone-or-a-tablet)

**Reply by flowerhorse** [reply]
> but you can do (live) [coding in GLSL like in ShaderToy](https://github.com/brunoherbelin/vimix/wiki/Filters-and-ShaderToy#custom-filter-with-shadertoy-coding)

Yes, I have already noticed that I have to deal with shadertoy as well.
Until now I am still with the program basics.

> Maybe this one ? Bzzzbz

What I am missing is an automatable visual/temporal element.
Something that you can run when you prepare something else.
The description of Bzzzbz reads very well from the possibilities.
Would milk-scripts also be possible?

> its [done with SRT](https://github.com/brunoherbelin/vimix/wiki/SRT-stream-I-O#srt-player-on-a-phone-or-a-tablet)

I'll try..

**Comment by brunoherbelin** [comment]
> What I am missing is an automatable visual/temporal element.
Something that you can run when you prepare something else.

Aha, here we are ! 
I have been thinking of this in different ways, but never ended fully satisfied with the UX concept.

The simpler idea is a sequencer, with a list of OSC commands that are executed one after the other, at the next beat or a given time. The whole sequence could loop. The programing of this is not difficult, but making the user interface for a list from top to bottom (or from left to right) is really boring and not something you do spontaneously. 
For this, I am thinking of implementing a pure text scripting, with a simple syntax `time; osc command [; osc command 2]`

    0.1; /vimix/0/alpha f 0.2f
    0.5; /vimix/0/alpha f 0.8f; /vimix/1/alpha f 0.3

A maybe a bit more interesting idea is an action recorder : hold the [SHIFT] key and everything you do with the mouse will be recorded: release [SHIFT] and your recording is now added to the action recorder. You can then play / pause / loop these recordings. But its not very visual and difficult to represent what the recording is about.

Finally, a more fun idea is the mechanical turk : a pseudo-machinery with pumps and wheels that are animated and move, and manipulate the sources by some attachment points  : e.g. a wheel turning in the mixing view, on which a source is attached to, so that the source would enter and exit the mixing circle (consequently rythmically fade in and out). This solution would be very visual, a bit fun and original, but i am afraid this would be anecdotical and in the end not very practical...

Glad to have your input on this!

**Comment by flowerhorse** [comment]
The first suggestion sounds good but too cryptic because of the command line.
If you could also read it in as a prepared batch file that would be great.

The completely visual solution I imagine spontaneously best because it needs less special knowledge.
But hard to program I think.
I will think about it when I deal with the keyboard inputs. Thematically it is similar.
I have to take a closer look there. Make some remarks about that soon.

**Comment by MoorPoor** [comment]
Hei hei!
+1 for some kind of video synthesizer! I can't contribute much to the topic at the moment, but I think the suggestion of an action recorder is basically a good idea. I would like to emphasize the power of the timeline.
for videos you have already integrated cropping and hiding - super simple and easy to use!

I imagine a timeline for each integrated medium.
if no runtime is stored, as in videos, a default value of e.g. 10 seconds (changeable) would makes sense?!
each value (of filters) could then be stored in this timeline and changed on the fly.

The workflow could look like this:
you click the medium e.g. Picture and in the timeline window the standard fade option opens. If I click on a filter, I can draw in the values there with the mouse or use Shift+slider movement in the settings to record them.

![Bildschirmfoto vom 2023-04-13 12-13-53](https://user-images.githubusercontent.com/60827766/231730010-499878a4-b392-48e3-ade9-ce2c88f96d70.png)

**Reply by brunoherbelin** [reply]
> I imagine a timeline for each integrated medium. if no runtime is stored, as in videos, a default value of e.g. 10 seconds (changeable) would makes sense?! each value (of filters) could then be stored in this timeline and changed on the fly.

yep, I guess it makes sense indeed.
Instead of by default, I imagine a button "+" to add a timeline on the play bar of a non-video source. 
The duration could be specified, and the same opacity and cut operations would be applied. 
I like it because it does not requires a new concept, it is already there.

> The workflow could look like this: you click the medium e.g. Picture and in the timeline window the standard fade option opens. If I click on a filter, I can draw in the values there with the mouse or use Shift+slider movement in the settings to record them.

here I might have lost you: do you mean that you imagine that the timeline curve (orange line) would modulate the parameter of the filter instead of the opacity of the source?

**Reply by MoorPoor** [reply]
> here I might have lost you: do you mean that you imagine that the timeline curve (orange line) would modulate the parameter of the filter instead of the opacity of the source?

not instead but additionally... so what about a timeline curve for each parameter as soon as this parameter has a changed value.
either you select the parameter in the filter section (e.g. by double-clicking), or in a dropdown next to the timeline?

![photo_2023-04-14_11-57-57](https://user-images.githubusercontent.com/60827766/232013386-a10ad925-541a-4af4-8359-98aa1e29442d.jpg)

**Comment by flowerhorse** [comment]
moved to separate Topic: Player Timeline BTW

**Comment by brunoherbelin** [comment]
Here is something that you should welcome ! 
I just implemented the possibility to add a timeline to a still image :

![image](https://github.com/brunoherbelin/vimix/assets/60393189/88f3f310-c478-4e1f-ad65-1b0400c8cde5)

Once you add a timeline and set its duration, the media behaves like a normal video.

**Reply by brunoherbelin** [reply]
NB: this is only in Beta 57154e5d0b74fc54829b64e629c2370f644db912 for now,

**Comment by brunoherbelin** [comment]
TImeline for images is now stable in vimix 0.8.1 (flatpak and snap)

![Screenshot from 2023-09-19 19-33-57](https://github.com/brunoherbelin/vimix/assets/60393189/90bee5e9-75de-441f-8060-10e85179af42)

Now, regarding the idea to control other attributes (e.g. gamma, brightness, etc) in the timeline, it is interesting but the timeline might not be where I would place it. We'll see how we can work on this !
